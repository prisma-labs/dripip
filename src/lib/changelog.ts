/**
 * This module deals with building changelogs from a serieses.
 */
import Chalk from 'chalk'
import { stripIndents } from 'common-tags'
import { Commit, Series } from '../utils/release'
import { casesHandled } from './utils'

type ChangeLog = {
  /**
   * Any commit marked as breaking.
   */
  breaking: {
    label: string
    commits: Commit[]
  }
  /**
   * Feat type commits.
   */
  features: {
    label: string
    commits: Commit[]
  }
  /**
   * Fix type commits.
   */
  fixes: {
    label: string
    commits: Commit[]
  }
  /**
   * Any type other than feat fix and chore.
   */
  improvements: {
    label: string
    commits: Commit[]
  }
  /**
   * Commits that are intended to be ignored.
   */
  chores: {
    label: string
    commits: Commit[]
  }
  /**
   * Commit mistakes, non-conforming tools, who knows. Shouldn't happen but can.
   */
  unspecified: {
    label: string
    commits: Commit[]
  }
}

export function empty(): ChangeLog {
  return {
    breaking: { commits: [], label: 'BREAKING CHANGES' },
    features: { commits: [], label: 'Features' },
    chores: { commits: [], label: 'Chores' },
    fixes: { commits: [], label: 'Fixes' },
    improvements: { commits: [], label: 'Improvements' },
    unspecified: { commits: [], label: 'Unspecified Changes' },
  }
}

function organize(series: Series): ChangeLog {
  const log = empty()

  for (const c of series.commitsInNextStable) {
    if (c.message.parsed === null) {
      log.unspecified.commits.push(c)
      continue
    }

    const cp = c.message.parsed

    // breaking changes are collected as a group in addition to by type.
    if (cp.breakingChange) {
      log.breaking.commits.push(c)
    }

    if (cp.typeKind === 'feat') {
      log.features.commits.push(c)
    } else if (cp.typeKind === 'fix') {
      log.fixes.commits.push(c)
    } else if (cp.typeKind === 'chore') {
      log.chores.commits.push(c)
    } else if (cp.typeKind === 'other') {
      log.improvements.commits.push(c)
    } else {
      casesHandled(cp.typeKind)
    }
  }

  return log
}

export function render(
  series: Series,
  opts: { type: 'plain' | 'markdown' }
): string {
  if (opts.type === 'markdown') return Markdown.render(organize(series))
  if (opts.type === 'plain') return Plain.render(organize(series))
  casesHandled(opts.type)
}

namespace Plain {
  export function render(log: ChangeLog): string {
    const order: (keyof Omit<ChangeLog, 'unspecified'>)[] = [
      'breaking',
      'features',
      'fixes',
      'improvements',
      'chores',
    ]

    const doc = order
      .filter(sectionName => {
        return log[sectionName].commits.length > 0
      })
      .map(sectionName => {
        if (sectionName === 'breaking') {
          return (
            sectionTitle(log[sectionName].label) +
            '\n\n' +
            sectionCommits(log[sectionName].commits, { breaking: false }) +
            '\n'
          )
        }

        if (sectionName === 'improvements') {
          return (
            sectionTitle(log[sectionName].label) +
            '\n\n' +
            sectionCommits(log[sectionName].commits, { type: true }) +
            '\n'
          )
        }

        return (
          sectionTitle(log[sectionName].label) +
          '\n\n' +
          sectionCommits(log[sectionName].commits) +
          '\n'
        )
      })

    if (log.unspecified.commits.length) {
      doc.push(
        sectionTitle(log.unspecified.label) +
          '\n\n' +
          '  ' +
          log.unspecified.commits
            .map(c => `${Chalk.gray(shortSha(c))} ${c.message.raw}`)
            .join('\n  ') +
          '\n'
      )
    }

    return doc.join('\n')
  }

  function sectionCommits(cs: Commit[], opts?: CommitRenderOpts): string {
    return cs.map(c => sectionCommit(c, opts)).join('\n')
  }

  function sectionTitle(title: string): string {
    return Chalk.magenta(title)
  }

  type CommitRenderOpts = { type?: boolean; breaking?: boolean }

  function sectionCommit(c: Commit, opts?: CommitRenderOpts): string {
    const sha = Chalk.gray(shortSha(c))
    const type = opts?.type === true ? ' ' + c.message.parsed!.type + ':' : ''
    const description = ' ' + c.message.parsed!.description
    const breaking =
      opts?.breaking === false
        ? ''
        : c.message.parsed!.breakingChange
        ? Chalk.red(' (breaking)')
        : ''
    return `  ${sha}${breaking}${type}${description}`
  }
}

namespace Markdown {
  export function render(log: ChangeLog): string {
    const order: (keyof Omit<ChangeLog, 'unspecified'>)[] = [
      'breaking',
      'features',
      'fixes',
      'improvements',
      'chores',
    ]

    const doc = order
      .filter(sectionName => {
        return log[sectionName].commits.length > 0
      })
      .map(sectionName => {
        if (sectionName === 'breaking') {
          return (
            stripIndents`
            ${sectionTitle(log[sectionName].label)}
  
            ${log[sectionName].commits
              .map(c => sectionCommit(c, { breaking: false }))
              .join('\n')}
          ` + '\n'
          )
        }

        if (sectionName === 'improvements') {
          return (
            stripIndents`
            ${sectionTitle(log[sectionName].label)}
  
            ${log[sectionName].commits
              .map(c => sectionCommit(c, { type: true }))
              .join('\n')}
          ` + '\n'
          )
        }

        return (
          stripIndents`
          ${sectionTitle(log[sectionName].label)}

          ${sectionCommits(log[sectionName].commits)}
        ` + '\n'
        )
      })

    if (log.unspecified.commits.length) {
      doc.push(
        stripIndents`
        ${sectionTitle(log.unspecified.label)}

        - ${log.unspecified.commits
          .map(c => `${shortSha(c)} ${c.message.raw}`)
          .join('\n- ')}
      ` + '\n'
      )
    }

    return doc.join('\n')
  }
  function sectionCommits(cs: Commit[]): string {
    return cs.map(c => sectionCommit(c)).join('\n')
  }

  function sectionTitle(title: string): string {
    return `#### ${title}`
  }

  function sectionCommit(
    c: Commit,
    opts?: { type?: boolean; breaking?: boolean }
  ): string {
    const sha = shortSha(c)
    const type = opts?.type === true ? ' ' + c.message.parsed!.type + ':' : ''
    const description = ' ' + c.message.parsed!.description
    const breaking =
      opts?.breaking === false
        ? ''
        : c.message.parsed!.breakingChange
        ? ' (breaking)'
        : ''
    return `- ${sha}${breaking}${type}${description}`
  }
}

function shortSha(c: Commit): string {
  return c.sha.slice(0, 7)
}
