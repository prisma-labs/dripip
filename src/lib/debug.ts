import createDebug from 'debug'

const dripipDebug = createDebug('dripip')

export function debug(formatter: any, ...args: any[]): void {
  return dripipDebug(formatter, ...args)
}
