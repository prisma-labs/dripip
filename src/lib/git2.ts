import * as nodefs from 'fs'
import isogit from 'isomorphic-git'
import http from 'isomorphic-git/http/node'

interface Input {
  cwd?: string
}

export type GitSyncStatus = 'synced' | 'not_synced' | 'remote_needs_branch'

class Git2 {
  private dir: string
  private fs: typeof nodefs
  private http: typeof http
  constructor(input?: Input) {
    this.dir = input?.cwd ?? process.cwd()
    this.fs = nodefs
    this.http = http
  }
  /**
   * Get the name of the currently checked out branch.
   */
  getCurrentBranchName() {
    return isogit.currentBranch({ fs: this.fs, dir: this.dir })
  }
  /**
   * Check how the local branch is not in sync or is with the remote.
   * Ref: https://stackoverflow.com/questions/3258243/check-if-pull-needed-in-git
   */
  async checkSyncStatus(input: { branchName: string }): Promise<GitSyncStatus> {
    let remoteUrl: string = await isogit.getConfig({
      fs: this.fs,
      dir: this.dir,
      path: 'remote.origin.url',
    })

    if (remoteUrl.startsWith('git@github.com:')) {
      remoteUrl = remoteUrl.replace('git@github.com:', 'https://github.com/')
    }
    if (!remoteUrl.endsWith('.git')) {
      remoteUrl = remoteUrl + '.git'
    }

    let remoteInfo

    try {
      remoteInfo = await isogit.getRemoteInfo({
        http: this.http,
        url: remoteUrl,
      })
    } catch (e) {
      throw new Error(`Failed to fetch remote info from ${remoteUrl} due to error:\n\n${e}`)
    }

    if (!remoteInfo.refs) {
      throw new Error('Could not fetch refs')
    }

    if (!remoteInfo.refs.heads) {
      throw new Error('Could not fetch ref heads')
    }

    if (
      !Object.keys(remoteInfo.refs.heads).find((remoteBranchName) => remoteBranchName === input.branchName)
    ) {
      return 'remote_needs_branch'
    }

    const localBranchHeadSha = await isogit.resolveRef({ fs: this.fs, dir: this.dir, ref: 'HEAD' })
    const remoteBranchHeadSha = remoteInfo.refs.heads[input.branchName]

    if (localBranchHeadSha === remoteBranchHeadSha) {
      return 'synced'
    }

    return 'not_synced'

    // todo https://github.com/isomorphic-git/isomorphic-git/issues/1110
  }
}

export function createGit(input?: Input) {
  return new Git2(input)
}
