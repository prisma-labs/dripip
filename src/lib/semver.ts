import * as SemVer from 'semver'

export type Ver = StableVer | PreviewVer

export type StableVer = {
  version: string
  vprefix: boolean
  major: string
  minor: string
  patch: string
}

export type PreviewVer = {
  version: string
  vprefix: boolean
  major: string
  minor: string
  patch: string
  preRelease: {
    identifier: string
    buildNum: number
  }
}

/**
 * Given a list of conventional commit messages (subject and body, the entire
 * message for the commit) calculate what the package version containing these
 * changes should be. Returns `null` if all changes were meta or unconforming.
 */
export function calcBumpType(
  commitMessages: string[]
): null | SemverStableVerParts {
  let semverPart: null | SemverStableVerParts = null
  for (const m of commitMessages) {
    // Commits that do not conform to conventional commit standard are discarded
    if (!isValidConventionalCommit(m)) {
      continue
    }

    // chore type commits are considered to not change the runtime in any way
    if (isMetaChange(m)) {
      continue
    }

    // Nothing can be be higher so we've reached our final value effectively.
    if (isBreakingchange(m)) {
      semverPart = 'major'
      break
    }

    // If already at minor continue, now looking only for major changes
    if (semverPart === 'minor') {
      continue
    }

    if (isMinorChange(m)) {
      semverPart = 'minor'
      continue
    }

    semverPart = 'patch'
  }

  return semverPart
}

function isMinorChange(message: string): boolean {
  return message.match(/^(?:feat|feature): /) !== null
}

function isBreakingchange(message: string): boolean {
  return message.match(/BREAKING CHANGE/) !== null
}

function isMetaChange(message: string): boolean {
  return message.match(/^chore: /) !== null
}

function isValidConventionalCommit(message: string): boolean {
  return message.match(/\w+: .+/) !== null
}

export type SemverStableVerParts = 'major' | 'minor' | 'patch'

/**
 * Calculate the stable bump to a given semver version. This function is similar
 * to `semver` package inc function with the following differences:
 *
 *     1. pre-releases are 1 based:
 *
 *          this  : '0.0.1'  inc('prerelease') --> '0.0.1-1'
 *          semver: '0.0.1'  inc('prerelease') --> '0.0.1-0'
 *
 *     2. bumping pre{min,maj,pat} also bumps the build num:
 *
 *          this  : '0.0.1-1' inc('preminor') --> '0.1.0-2'
 *          semver: '0.0.1'   inc('preminor') --> '0.1.0-0'
 */
export function bump(
  bumpType: 'major' | 'minor' | 'patch',
  // | 'premajor'
  // | 'preminor'
  // | 'prepatch'
  // | 'pre',
  // preReleaseTypeIdentifier: string,
  prevVer: SemVer.SemVer
): SemVer.SemVer {
  // const buildNumPrefix = preReleaseTypeIdentifier
  //   ? `${preReleaseTypeIdentifier}.`
  //   : ''
  switch (bumpType) {
    case 'major':
      return SemVer.parse(`${prevVer.major + 1}.0.0`)!
    case 'minor':
      return SemVer.parse(`${prevVer.major}.${prevVer.minor + 1}.0`)!
    case 'patch':
      return SemVer.parse(
        `${prevVer.major}.${prevVer.minor}.${prevVer.patch + 1}`
      )!
    // // TODO refactor
    // case 'premajor':
    //   // TODO unsafe, assumes the incoming ver has format #.#.# or #.#.#-foo.#
    //   const buildNum1 = (prevVer.prerelease[1] as undefined | number) ?? 1
    //   const preRelease1 = buildNumPrefix + String(buildNum1 + 1)
    //   return Semver.parse(
    //     `${prevVer.major + 1}.${prevVer.minor}.${prevVer.patch}-${preRelease1}`
    //   )!
    // case 'preminor':
    //   // TODO unsafe, assumes the incoming ver has format #.#.# or #.#.#-foo.#
    //   const buildNum2 = (prevVer.prerelease[1] as undefined | number) ?? 1
    //   const preRelease2 = buildNumPrefix + String(buildNum2 + 1)
    //   return Semver.parse(
    //     `${prevVer.major}.${prevVer.minor + 1}.${prevVer.patch}-${preRelease2}`
    //   )!
    // case 'prepatch':
    //   // TODO unsafe, assumes the incoming ver has format #.#.# or #.#.#-foo.#
    //   const buildNum3 = (prevVer.prerelease[1] as undefined | number) ?? 1
    //   const preRelease3 = buildNumPrefix + String(buildNum3 + 1)
    //   return Semver.parse(
    //     `${prevVer.major}.${prevVer.minor}.${prevVer.patch + 1}-${preRelease3}`
    //   )!
    // case 'pre':
    //   // TODO unsafe, assumes the incoming ver has format #.#.# or #.#.#-foo.#
    //   const buildNum4 = (prevVer.prerelease[1] as undefined | number) ?? 1
    //   const preRelease4 = buildNumPrefix + String(buildNum4 + 1)
    //   return Semver.parse(
    //     `${prevVer.major}.${prevVer.minor}.${prevVer.patch}-${preRelease4}`
    //   )!
  }
}

/**
 * Create a semver instance programatically.
 */
export const create = (maj: number, min: number, pat: number) =>
  SemVer.parse(`${maj}.${min}.${pat}`)!

export function parsePreview(ver: string): null | PreviewVer {
  const result = parse(ver)
  if (result === null) return null
  if ((result as any).preRelease) {
    return result as PreviewVer
  }
  throw new Error(
    `Given version string ${ver} could not be parsed as a preview.`
  )
}

/**
 * Parse a version string into structured data.
 */
export function parse(ver: string): null | StableVer | PreviewVer {
  const result = ver.match(
    /^(v)?(\d+).(\d+).(\d+)$|^(v)?(\d+).(\d+).(\d+)-(\w+).(\d+)$/
  )

  if (result === null) return null

  if (result[6]) {
    const vprefix = result[5] === 'v'
    const major = result[6]
    const minor = result[7]
    const patch = result[8]
    const identifier = result[9]
    const buildNum = parseInt(result[10], 10)
    return {
      version: `${major}.${minor}.${patch}-${identifier}.${buildNum}`,
      vprefix,
      major,
      minor,
      patch,
      preRelease: {
        identifier,
        buildNum,
      },
    }
  }

  const vprefix = result[1] === 'v'
  const major = result[2]
  const minor = result[3]
  const patch = result[4]
  return {
    version: `${major}.${minor}.${patch}`,
    vprefix,
    major,
    minor,
    patch,
  }
}

export const parseToClass = SemVer.parse

export { SemVer } from 'semver'
