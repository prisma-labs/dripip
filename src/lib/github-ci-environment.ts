import { rootDebug } from './debug'

const debug = rootDebug(__filename)

export interface GitHubCIEnvironment {
  runId: number
  eventName: 'pull_request'
  ref: null | string
  headRef: null | string
  repository: string
  parsed: {
    repo: {
      name: string
      owner: string
    }
    branchName: null | string
    prNum?: number
  }
}

/**
 * Parse the GitHub CI Environment. Returns null if parsing fails which should
 * mean it is not a GitHub CI Environment.
 *
 * @remarks
 *
 * GitHub docs: https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables
 */
export function parseGitHubCIEnvironment(): null | GitHubCIEnvironment {
  if (!isGitHubCIEnvironment()) return null

  let prNum: GitHubCIEnvironment['parsed']['prNum']

  if (process.env.GITHUB_REF) {
    const match = process.env.GITHUB_REF.match(/refs\/pull\/(\d+)\/merge/)

    if (match) {
      debug('found a pr number from github ci environment %s', match[1])
      prNum = parseInt(match[1], 10)
    }
  }

  const repoPath = process.env.GITHUB_REPOSITORY!.split('/')

  const repo = {
    owner: repoPath[0],
    name: repoPath[1],
  }

  let branchName: null | string = null
  if (process.env.GITHUB_HEAD_REF) {
    branchName = process.env.GITHUB_HEAD_REF
  } else if (process.env.GITHUB_REF) {
    branchName = process.env.GITHUB_REF.match(/^refs\/heads\/(.+)$/)?.[1] ?? null
  }

  return {
    runId: parseInt(process.env.GITHUB_RUN_ID!, 10),
    eventName: process.env.GITHUB_EVENT_NAME! as GitHubCIEnvironment['eventName'],
    ref: process.env.GITHUB_REF ?? null,
    headRef: process.env.GITHUB_HEAD_REF ?? null,
    repository: process.env.GITHUB_REPOSITORY!,
    parsed: {
      prNum: prNum,
      repo: repo,
      branchName: branchName,
    },
  }
}

/**
 * Check if the current process appears to be running in a GitHub CI environment.
 */
export function isGitHubCIEnvironment() {
  return process.env.GITHUB_RUN_ID !== undefined
}
