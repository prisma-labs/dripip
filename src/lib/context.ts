import createGit from 'simple-git/promise'
import Octokit from '@octokit/rest'
import * as Git from './git'
import {
  getReleasesAtCommit,
  isStable,
  isStablePreview,
  findLatestStable,
  findLatestPreview,
} from './utils'
import * as SemVer from 'semver'

export type Context = {
  commitsSinceLastStable: Git.LogEntry[]
  commitsSinceLastPreview: Git.LogEntry[]
  // nextReleasesNowWouldBe: {
  //   stable: null | SemVer.SemVer
  //   preview: null | SemVer.SemVer
  // }
  latestReleases: {
    stable: null | StableRelease
    preview: null | PreviewRelease
  }
  githubRepo: {
    owner: string
    name: string
    trunkBranch: string
  }
  currentBranch: {
    name: string
    isDetatched: boolean
    isTrunk: boolean
    syncStatus: Git.SyncStatus
    prs: {
      open: null | PullRequest
      closed: PullRequest[]
    }
  }
  currentCommit: {
    sha: string
    releases: {
      stable: null | StableRelease
      preview: null | PreviewRelease
    }
  }
}

type Commit = {
  sha: string
  message: string
}

type PullRequest = {
  title: string
  number: number
}

type StableRelease = {
  type: 'stable'
  version: SemVer.SemVer
  sha: string
}

type PreviewRelease = {
  type: 'preview'
  version: SemVer.SemVer
  sha: string
}

export async function scan(): Promise<Context> {
  // Build up instances
  const octoOps = {} as Octokit.Options
  if (process.env.GITHUB_TOKEN) octoOps.auth = process.env.GITHUB_TOKEN
  const octokit = new Octokit(octoOps)
  const git = createGit()
  // Get the trunk branch
  const githubRepoAddress = await Git.parseGithubRepoInfoFromGitConfig()
  const githubRepo = await octokit.repos.get({
    owner: githubRepoAddress.owner,
    repo: githubRepoAddress.name,
  })
  const trunkBranch = githubRepo.data.default_branch
  const branchesSummary = await git.branch({})
  // Get the PRs if any
  const prs = (
    await octokit.pulls.list({
      owner: githubRepoAddress.owner,
      repo: githubRepoAddress.name,
      head: `${githubRepoAddress.owner}:${branchesSummary.current}`,
      state: 'all',
    })
  ).data
  // Github only permits one open PR per branch name. If there is an open one
  // then we can discard any others as inconsequential. If none are open, and
  // there are multiple, we should not discard them, as they may be of interest
  // to the user to debug an exception, e.g. attempting a pr release without a
  // pr open: then we can say "Hey, you have no pr open, but oddly we did find
  // these past PRs...?"
  const prsByState: Context['currentBranch']['prs'] = {
    open: null,
    closed: [],
  }
  for (const pr of prs) {
    if (pr.state === 'open') {
      prsByState.open = { number: pr.number, title: pr.title }
    } else {
      prsByState.closed.push({ number: pr.number, title: pr.title })
    }
  }
  // get the branch sync status
  const syncStatus = await Git.checkSyncStatus(git)
  // get commit info
  const currentCommitSHA = await Git.gitGetSha(git, { ref: 'head' })
  const commitReleases = await getReleasesAtCommit(currentCommitSHA)
  // get the latest releases
  const maybeLatestStableVer = await findLatestStable(git)
  const maybeLatestPreVerSinceStable = await findLatestPreview(
    git,
    maybeLatestStableVer
  )
  // get commits since the latest releases
  const [commitsSinceLastStable, commitsSinceLastPreview] = await Promise.all([
    Git.log(git, { since: maybeLatestStableVer }),
    Git.log(git, { since: maybeLatestPreVerSinceStable }),
  ])

  // return the final result
  return {
    commitsSinceLastPreview,
    commitsSinceLastStable,
    latestReleases: {
      stable:
        maybeLatestStableVer === null
          ? null
          : {
              type: 'stable',
              sha: 'todo',
              version: SemVer.parse(maybeLatestStableVer)!,
            },
      preview:
        maybeLatestPreVerSinceStable === null
          ? null
          : {
              type: 'preview',
              sha: 'todo',
              version: SemVer.parse(maybeLatestPreVerSinceStable)!,
            },
    },
    githubRepo: {
      owner: githubRepoAddress.owner,
      name: githubRepoAddress.name,
      trunkBranch,
    },
    currentBranch: {
      name: branchesSummary.current,
      isTrunk: branchesSummary.current === trunkBranch,
      isDetatched: branchesSummary.detached,
      prs: prsByState,
      syncStatus,
    },
    currentCommit: {
      sha: currentCommitSHA,
      releases: commitReleases,
    },
  }
}
