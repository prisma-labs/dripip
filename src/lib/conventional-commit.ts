import * as Semver from './semver'

/**
 * Given a list of conventional commit messages (subject and body, the entire
 * message for the commit) calculate what the package version containing these
 * changes should be. Returns `null` if all changes were meta or unconforming.
 */
export function calcBumpType(
  isInitialDevelopment: boolean,
  commitMessages: string[]
): null | Semver.MajMinPat {
  let semverPart: null | Semver.MajMinPat = null
  for (const m of commitMessages) {
    const cc = parse(m)

    // Commits that do not conform to conventional commit standard are discarded
    if (!cc) continue

    // Completing initial development is a spec extension
    // https://github.com/conventional-commits/conventionalcommits.org/pull/214
    if (isInitialDevelopment && cc.completesInitialDevelopment) return 'major'

    if (isMetaChange(cc)) {
      // chore type commits are considered to not change the runtime in any way
      continue
    }

    // Nothing can be be higher so we've reached our final value effectively.
    if (cc.breakingChange) {
      // during initial development breaking changes are permitted without
      // having to bump the major.
      semverPart = isInitialDevelopment ? 'minor' : 'major'
      break
    }

    // If already at minor continue, now looking only for major changes
    if (semverPart === 'minor') {
      continue
    }

    if (isMinorChange(cc)) {
      semverPart = 'minor'
      // during initial development breaking changes are permitted without
      // having to bump the major. Therefore, we know we won't get a bumpType
      // higher than this, can short-circuit.
      if (isInitialDevelopment) break
      else continue
    }

    semverPart = 'patch'
  }

  return semverPart
}

function isMinorChange(conventionalCommit: ConventionalCommit): boolean {
  return ['feat', 'feature'].includes(conventionalCommit.type)
}

function isMetaChange(conventionalCommit: ConventionalCommit): boolean {
  return ['chore'].includes(conventionalCommit.type)
}

type Kind = 'feat' | 'fix' | 'chore' | 'other'

export type ConventionalCommit = {
  typeKind: Kind
  type: string
  scope: null | string
  description: string
  body: null | string
  breakingChange: null | string
  footers: { type: string; body: string }[]
  completesInitialDevelopment: boolean
}

const pattern = /^([^:\r\n(!]+)(?:\(([^\r\n()]+)\))?(!)?:\s*([^\r\n]+)[\n\r]*(.*)$/s

export function parse(message: string): null | ConventionalCommit {
  const result = message.match(pattern)
  if (!result) return null
  const [, type, scope, breakingChangeMarker, description, rest] = result

  let completesInitialDevelopment = false
  let breakingChange = breakingChangeMarker === undefined ? null : 'No Explanation'
  let body = null
  let footers: ConventionalCommit['footers'] = []

  if (rest) {
    const rawFooters: string[] = []

    let currFooter = -1
    let currSection = 'body'
    for (const para of rest.split(/(?:\r?\n){2}/)) {
      if (para.match(/^COMPLETES[-\s]INITIAL[-\s]DEVELOPMENT\s*/)) {
        completesInitialDevelopment = true
      } else if (para.match(/^\s*BREAKING[-\s]CHANGE\s*:\s*.*/)) {
        currSection = 'breaking_change'
        breakingChange = (breakingChange ?? '') + '\n\n' + para.replace(/^BREAKING[-\s]CHANGE\s*:/, '')
      } else if (para.match(/^\s*[\w-]+\s*:.*/)) {
        currSection = 'footers'
        rawFooters.push(para)
        currFooter++
      } else if (currSection === 'body') {
        body = (body ?? '') + '\n\n' + para
      } else if (currSection === 'breaking_change') {
        breakingChange = (breakingChange ?? '') + '\n\n' + para
      } else {
        rawFooters[currFooter] += '\n\n' + para
      }
    }

    footers = rawFooters.map((f) => {
      const [, type, body] = f.trim().split(/^\s*([\w-]+)\s*:/)
      return {
        type: type.trim(),
        body: body.trim(),
      }
    })
  }

  const typeTrimmed = type.trim()

  return {
    typeKind: getKind(typeTrimmed),
    type: typeTrimmed,
    scope: scope?.trim() ?? null,
    description: description.trim(),
    body: body?.trim() ?? null,
    footers: footers ?? [],
    breakingChange: breakingChange?.trim() ?? null,
    completesInitialDevelopment,
  }
}

function getKind(s: string): Kind {
  if (s === 'feat') return 'feat'
  if (s === 'fix') return 'fix'
  if (s === 'chore') return 'chore'
  return 'other'
}
