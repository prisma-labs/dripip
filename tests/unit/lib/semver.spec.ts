import * as Semver from '../../../src/lib/semver'

describe('parse', () => {
  it('0.0.1', () => {
    expect(Semver.parse('0.0.1')).toMatchSnapshot()
  })

  it('0.1.0', () => {
    expect(Semver.parse('0.1.0')).toMatchSnapshot()
  })

  it('1.0.0', () => {
    expect(Semver.parse('1.0.0')).toMatchSnapshot()
  })

  it('0.0.11', () => {
    expect(Semver.parse('0.0.11')).toMatchSnapshot()
  })

  it('0.11.0', () => {
    expect(Semver.parse('0.11.0')).toMatchSnapshot()
  })

  it('11.0.0', () => {
    expect(Semver.parse('11.0.0')).toMatchSnapshot()
  })

  describe('v-prefix', () => {
    it('v0.0.0', () => {
      expect(Semver.parse('v0.0.0')).toMatchSnapshot()
    })

    it('v0.0.0-a.1', () => {
      expect(Semver.parse('v0.0.0-a.1')).toMatchSnapshot()
    })
  })

  describe('pre-releases', () => {
    it('0.0.0-a.1', () => {
      expect(Semver.parse('0.0.0-a.1')).toMatchSnapshot()
    })
  })

  describe('rejects', () => {
    it.each([
      ['1'],
      [''],
      ['a.b.c'],
      ['0.0.0-'],
      ['0.0.0-1'],
      ['0.0.0-a'],
      ['0.0.0-a.a'],
    ])('%s', caze => {
      expect(Semver.parse(caze)).toBeNull()
    })
  })
})

describe('calcBumpType', () => {
  it('"fix" bumps patch', () => {
    expect(Semver.calcBumpType(['fix: 1'])).toEqual('patch')
  })

  it('"feat" bumps minor', () => {
    expect(Semver.calcBumpType(['feat: 1'])).toEqual('minor')
  })

  it('"feature" bumps minor', () => {
    expect(Semver.calcBumpType(['feat: 1'])).toEqual('minor')
  })

  it('presence of "BREAKING CHANGE" bumps major', () => {
    expect(
      Semver.calcBumpType(['anything: 1\nBREAKING CHANGE\nfoobar'])
    ).toEqual('major')
  })

  it('an unknown change type bumps patch', () => {
    expect(Semver.calcBumpType(['anything: 1'])).toEqual('patch')
  })

  it('patch-level changes ignored if already bumped past patch', () => {
    expect(Semver.calcBumpType(['feat: 1', 'fix: 1'])).toEqual('minor')
    expect(Semver.calcBumpType(['fix: 1', 'feat: 1'])).toEqual('minor')
  })

  it('feat-level changes ignored if already bumped past minor', () => {
    expect(Semver.calcBumpType(['feat: 1', 'fix: 1\nBREAKING CHANGE'])).toEqual(
      'major'
    )
    expect(Semver.calcBumpType(['fix: 1\nBREAKING CHANGE', 'feat: 1'])).toEqual(
      'major'
    )
  })

  it('chore-type commits are ignored', () => {
    expect(Semver.calcBumpType(['chore: 1'])).toEqual(null)
    expect(Semver.calcBumpType(['chore: 1', 'fix: 1'])).toEqual('patch')
    expect(Semver.calcBumpType(['chore: 1', 'feat: 1'])).toEqual('minor')
    expect(Semver.calcBumpType(['chore: 1\nBREAKING CHANGE\nfoobar'])).toEqual(
      null
    )
    expect(
      Semver.calcBumpType(['chore: 1', 'fix: 1\nBREAKING CHANGE'])
    ).toEqual('major')
  })

  it('invalid message formats cause the commit to be ignored', () => {
    expect(Semver.calcBumpType(['unknown'])).toEqual(null)
    expect(Semver.calcBumpType(['unknown', 'fix: 1'])).toEqual('patch')
    expect(Semver.calcBumpType(['unknown', 'feat: 1'])).toEqual('minor')
    expect(Semver.calcBumpType(['unknown\nBREAKING CHANGE\nfoobar'])).toEqual(
      null
    )
    expect(Semver.calcBumpType(['unknown', 'fix: 1\nBREAKING CHANGE'])).toEqual(
      'major'
    )
  })
})

describe('bumpVer', () => {
  it('can bump patch', () => {
    expect(Semver.bump('major', Semver.create(1, 1, 1)).version).toEqual(
      '2.0.0'
    )
  })
  it('can bump minor', () => {
    expect(Semver.bump('minor', Semver.create(1, 1, 1)).version).toEqual(
      '1.2.0'
    )
  })
  it('can bump major', () => {
    expect(Semver.bump('major', Semver.create(1, 1, 1)).version).toEqual(
      '2.0.0'
    )
  })
})
