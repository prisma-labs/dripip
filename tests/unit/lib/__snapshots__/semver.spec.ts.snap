// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 0.0.1 1`] = `
Object {
  "major": "0",
  "minor": "0",
  "patch": "1",
  "version": "0.0.1",
  "vprefix": false,
}
`;

exports[`parse 0.0.11 1`] = `
Object {
  "major": "0",
  "minor": "0",
  "patch": "11",
  "version": "0.0.11",
  "vprefix": false,
}
`;

exports[`parse 0.1.0 1`] = `
Object {
  "major": "0",
  "minor": "1",
  "patch": "0",
  "version": "0.1.0",
  "vprefix": false,
}
`;

exports[`parse 0.11.0 1`] = `
Object {
  "major": "0",
  "minor": "11",
  "patch": "0",
  "version": "0.11.0",
  "vprefix": false,
}
`;

exports[`parse 1.0.0 1`] = `
Object {
  "major": "1",
  "minor": "0",
  "patch": "0",
  "version": "1.0.0",
  "vprefix": false,
}
`;

exports[`parse 11.0.0 1`] = `
Object {
  "major": "11",
  "minor": "0",
  "patch": "0",
  "version": "11.0.0",
  "vprefix": false,
}
`;

exports[`parse pre-releases 0.0.0-a.1 1`] = `
Object {
  "major": "0",
  "minor": "0",
  "patch": "0",
  "preRelease": Object {
    "buildNum": 1,
    "identifier": "a",
  },
  "version": "0.0.0-a.1",
  "vprefix": false,
}
`;

exports[`parse v-prefix v0.0.0 1`] = `
Object {
  "major": "0",
  "minor": "0",
  "patch": "0",
  "version": "0.0.0",
  "vprefix": true,
}
`;

exports[`parse v-prefix v0.0.0-a.1 1`] = `
Object {
  "major": "0",
  "minor": "0",
  "patch": "0",
  "preRelease": Object {
    "buildNum": 1,
    "identifier": "a",
  },
  "version": "0.0.0-a.1",
  "vprefix": true,
}
`;
